aten/src/ATen/native/DilatedConvolution.cpp

```diff
@@ -0,0 +1,879 @@
+  int dimh = dim - 1;
+  int dimw = dim;
+
+  if (ndim == dim + 2) {
+    dimd++;
+    dimf++;
+    dimh++;
+    dimw++;
+  }
+  TORCH_CHECK(
+      (ndim == dim + 2 || ndim == dim + 1),
+      "non-empty ",
+      dim + 1,
+      "D or ",
+      dim + 2,
+      "D input tensor expected but got: ",
+      input.sizes());
+
+  switch (dim) {
+    case 2: {
+      int64_t inputHeight = input.size(dimh);
+      int64_t inputWidth = input.size(dimw);
+      int64_t outputHeight = OUTPUTSIZE(0, dim);
+      int64_t outputWidth = OUTPUTSIZE(1, dim);
+      TORCH_CHECK(
+          outputWidth >= 0 && outputHeight >= 0,
+          "Given input size per channel: (",
+          inputHeight,
+          " x ",
+          inputWidth,
+          "). "
+          "Calculated output size per channel: (",
+          outputHeight,
+          " x ",
+          outputWidth,
+          "). Output size is too small");
+      if (grad_output.defined()) {
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimh, outputHeight);
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimw, outputWidth);
+      }
+    } break;
+    case 3: {
+      int64_t inputDepth = input.size(dimd);
+      int64_t inputHeight = input.size(dimh);
+      int64_t inputWidth = input.size(dimw);
+      int64_t outputDepth = OUTPUTSIZE(0, dim);
+      int64_t outputHeight = OUTPUTSIZE(1, dim);
+      int64_t outputWidth = OUTPUTSIZE(2, dim);
+      TORCH_CHECK(
+          outputWidth >= 0 && outputHeight >= 0 && outputDepth >= 0,
+          "Given input size per channel: (",
+          inputDepth,
+          " x ",
+          inputHeight,
+          " x ",
+          inputWidth,
+          "). "
+          "Calculated output size per channel: (",
+          outputDepth,
+          " x ",
+          outputHeight,
+          " x ",
+          outputWidth,
+          "). Output size is too small");
+      if (grad_output.defined()) {
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimd, outputDepth);
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimh, outputHeight);
+        TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimw, outputWidth);
+      }
+    } break;
+    default:
+      TORCH_CHECK(false, "unexpected dim in conv_dilated shape check", dim);
+  } // switch
+
+  TORCH_CHECK_DIM_SIZE(input, ndim, dimf, weight.size(1));
+
+  if (grad_output.defined()) {
+    TORCH_CHECK(
+        grad_output.is_contiguous(), "grad_output ought to be contiguous");
+    TORCH_CHECK_DIM_SIZE(grad_output, ndim, dimf, weight.size(0));
+  }
+
+  if (grad_input.defined() || grad_weight.defined() || grad_bias.defined()) {
+    TORCH_CHECK(
+        grad_output.defined(), "grad_output must be defined for gradients");
+  }
+
+  if (columns.defined()) {
+    TORCH_CHECK(columns.is_contiguous(), "columns needs to be contiguous");
+  }
+
+} // conv_dilated_shape_check
+
+/*
+  conv_dilated2d_all_cpu_template
+
+  Main worker. Computes tensors output, grad_input, grad_weight,
+  and/or grad_bias if non-empty.
+ */
```

Comment:
To be honest, it's kind of confusing now that both forwards and backwards has been smooshed into the same function.

Reply:
I joined these because of the forward and backward function loops have the same structure: local variables, argument checking, loops. Splitting it up is certainly possible if wished.

Reply:
I hope that we can keep these in one function as now also 2d and 3d versions of the templates are joined into one C++ template functions. Code reduction is considerable.

If at all realistic, the conv_dilated functions could be generalized to arbitrary dimensions, one would just have to provide hyperplane2col and col2hyperplane implementations for that.